### DISCUSSION TITLE
음파 생성에 관한 성능 개선방안

### DISCUSSION THREAD

**Discussion by jopopcorn** [euphony-io/check-in-speaker#2] [General]
@euphony-io/euphoria 
얼마 전에 Euphony 라이브러리의 0.8 버전이 업데이트 되었습니다.
Tx에 추가된 인터페이스와 클래스들을 확인하고, 어떤 기능을 제공하는지 우선 분석해보는 게 좋을 것 같아요.
코드를 읽어보면서 궁금한 사항이나 캐치할 부분이 보이면 자유롭게 코멘트 남겨주세요! 😸

**Comment by designe** [comment]
음파 발생을 담당하다보니 ASK 피쳐를 한번 만들어보는 것도 향후 도움되지 않을까 싶어요! :)

**Reply by kmsaidev** [reply]
네!! 일단 ASK 피쳐에 대해서 공부해보고 정리해서 discussion에 공유하겠습니다!!

**Reply by designe** [reply]
좋아여! 음파 발생쪽은 누구누구가 담당하시나요?!

**Reply by jopopcorn** [reply]
> 좋아여! 음파 발생쪽은 누구누구가 담당하시나요?!

우선은 앱 개발을 서포트로 하시는 미소님, 하은님, 호영님을 중점으로 담당하기로 했습니다!
앱 개발은 주말내로 끝내고 최대한 다같이 참여해서 ASK 개발과 성능 개선에 기여해보면 좋을 것 같습니다!

**Reply by yeonns** [reply]
@kmsaidev @hoho00 @HACore 혹시 공부하시다가 어렵거나 궁금하신게 있으면 디스커션에 남겨주세요! 도움드리겠습니다 😄

**Reply by kmsaidev** [reply]
> @kmsaidev @hoho00 @HACore 혹시 공부하시다가 어렵거나 궁금하신게 있으면 디스커션에 남겨주세요! 도움드리겠습니다 😄 

넵 감사합니다!!😀😀

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 
다들 공부 잘 하고 계시나요?! 저는 euphony의 기술동향 보고서를 다시 읽어보았는데요.
유포니 라이브러리 아키텍처 구조와 0.8.0 버전의 전반적인 특징 등이 자세히 나와 있어서 여러분도 다시 읽어보면 좋을 것 같습니다!

지웅 멘토님의 블로그 포스팅 중 하나인 [통신 모듈레이션 방식에 관한 연구 및 euphony.js v0.2.0 release](https://blog.jbear.co/post/euphony_js_0.2.0/)라는 글도 참고하시면 좋을 것 같아요!!
모듈레이션에 관한 이야기가 나와 있어서 아직 배경지식이 부족하다면 이해하기 좋을 것 같습니다!

**Reply by psh2849** [reply]
정보 감사합니다 ㅎㅎ

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 
기술동향 보고서에 나오는 유포니 라이브러리의 아키텍처를 보시면 아시겠지만, 모든 음파 처리 부분을 네이티브로 구현하므로 c++로 작업을 해야 합니다!
통신에 관한 지식이 부족한 것도 있지만 c++을 아예 몰라서 코드를 이해하는데 많은 어려움이 있네요 ㅜㅜ
FSK 코드부터 살펴보고 있는데 우선 이해한 흐름대로 적어보려고 합니다!

</br>

FSK는 디지털 변조 방식 중의 하나이며, 주파수를 기준으로 데이터를 구분할 수 있습니다.
FSK를 쉽게 이해하기 위해 예시를 들자면 0과 1을 표현할 때 1초 동안의 주파수가 크다면 1, 적으면 0으로 표현할 수 있습니다.

</br>

FSK 헤더 파일에는 다음과 같이 정의되어 있습니다.
'Modem이 디지털 변조 방식이다'라고 생각하면 좋을 것 같고, FSK가 그러한 디지털 변조 방식 중 하나입니다.

`modulate` 함수가 두가지인데, 아래 빨간색 동그라미로 표시한 버튼을 누르면 내부 코드를 따라가 확인할 수 있습니다.
버튼을 누르면 `FSK.cpp` 파일에서의 `modulate` 구현체를 볼 수 있습니다.

<img width="1057" alt="스크린샷 2021-10-08 오전 2 42 22" src="https://user-images.githubusercontent.com/35393459/136436285-55623e0d-4a0b-48aa-9ef3-196992dd083f.png">

</br>

19라인의 `WaveList FSK::modulate(string code)` 함수의 코드를 먼저 보면, Base16 기반으로 문자를 아날로그(음파)로 만들어줍니다.
0부터 f까지의 문자를 각각의 음파로 변조해서 생성합니다. 그런데 25라인에는 S라는 문자도 케이스에 포함되어 있어서 뭐지?! 했습니다.

```c++
WaveList FSK::modulate(string code) {

    vector<shared_ptr<Wave>> result;

    for (char c : code ) {
        switch(c) {
            case 'S':
                result.push_back(
                        Wave::create()
                                .vibratesAt(kStartSignalFrequency)
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
        ...

```

</br>

`Packet.h` 파일을 열어보면 `toString()` 함수가 있는데, 위의 사진에 표시한 버튼과 마찬가지로 toString() 라인의 왼쪽 버튼을 클릭하면 코드 구현체가 나옵니다. 그냥 `Packet.cpp` 파일 들어가서 바로 확인해도 상관 없지만, 저는 보통 아래에 첨부한 사진에서 보이는 **Edit Source**라는 메뉴의 커맨드를 사용해서 내부 코드를 확인하는 편이라서 이런 흐름으로 본다는 것을 참고로 작성합니다!
<img width="720" alt="스크린샷 2021-10-08 오전 2 53 43" src="https://user-images.githubusercontent.com/35393459/136437665-ac1a5d12-990e-4617-94b3-775b4cec6f4d.png">
 
</br>

어쨌든 패킷을 String으로 바꿔줄 때 result 값에 S라는 문자를 추가하고 그 뒤에 페이로드와 체크섬, 패리티 코드를 뒤이어 붙이는 것을 확인했습니다.
멘토님이 올려주신 v0.8 변경사항을 보니 Packet 클래스는 음파 통신의 데이터 패킷을 의미한다고 합니다.
예상하는 시나리오로는 디지털 신호를 패킷으로 받아 String으로 변환하고, 각 문자에 맞는 음파로 변조하는 것 같습니다!
```c++
std::string Packet::toString() {
    std::stringstream result;

    result << "S" <<
    payload->getBaseString() <<
    checksum->getBaseString() <<
    parityCode->getBaseString();

    return result.str();
}
```

</br>

이제 `demodulate` 함수를 보려고 합니다! WaveList라는 Wave를 저장한 벡터의 사이즈 만큼 HexVector 변수를 생성해줍니다.
멘토님의 v0.8 변경사항 설명에 따르자면, 기본적으로 모든 Base는 HexVector를 가지고 각각의 진법으로 변환하여 사용한다고 합니다.
HexVector의 생성자 내부 코드를 보면 문자를 다시 16진법으로 변환하는 것으로 추측됩니다.

```c++
HexVector::HexVector(const std::string& hexString) {
    for(char c : hexString) {
        switch(c) {
            case '0': case '1': case '2':
            case '3': case '4': case '5':
            case '6': case '7': case '8':
            case '9': default:
                hexSource.push_back(c - '0');
                break;
            case 'a': case 'b': case 'c':
            case 'd': case 'e': case 'f':
                hexSource.push_back(10 + c - 'a');
                break;
        }
    }
}
```

</br>

for문에서는 WaveList를 순회하여 각 웨이브를 int형으로 전환한 후 스펙트럼을 만들어 HexVector 변수에 push 해줍니다.
이러한 HexVector를 패킷으로 바꾸어서 반환하는 것이 복조입니다.

```c++
shared_ptr<Packet> FSK::demodulate(const WaveList& waveList) {
    HexVector hexVector = HexVector(waveList.size());
    
    for(const auto& wave : waveList) {
        auto vectorInt16Source = wave->getInt16Source();
        int16_t* int16Source = &vectorInt16Source[0];
        float *resultBuf = fftModel->makeSpectrum(int16Source);
        hexVector.pushBack(getMaxIdxFromSource(resultBuf));
    }

    return std::make_shared<Packet>(hexVector);
}
```

</br>

있는 그대로의 코드를 당연하게 설명한 것 뿐이긴 하지만, 아직 이해를 못하셨다면 참고하시길 바랍니다!
제 해석이 정확하진 않으므로 읽고 정정해야 할 부분은 꼭 말씀해주세요 😹

**Reply by designe** [reply]
@jopopcorn 
이 내용은 다른 멘티들도 공유하면 너무 좋을 것 같네요!
정리해서 euphony discussion에 올려보는거 어떨까요?

**Reply by jopopcorn** [reply]
> @jopopcorn 이 내용은 다른 멘티들도 공유하면 너무 좋을 것 같네요! 정리해서 euphony discussion에 올려보는거 어떨까요?

알겠습니다! 좀 더 정리해서 올려보도록 하겠습니다!
